proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,351 @@
+      "16": darkImageData16,
+      "32": darkImageData32,
+      "color_scheme": "dark"
+    }, {
+      "16": lightImageData16,
+      "32": lightImageData32,
+      "color_scheme": "light"
+    }
+  ]
+});
+```
+
+* Pros: Future resistant allowing for more keys such as density (e.g. 2dppx),
+purpose (e.g. monochrome), and etc.
+* Cons: New paradigm for icon definition, and unclear how to handle color_schemes.
+* Questions: Should `color_scheme` be used instead of `mode`?
+---
+
+#### Option (pathVariant option)
+manifest.json from the combined option.
+
+action.setIcon()
+```
+action.setIcon({
+  pathVariants,
+  imageDataVariants,
+});
+
+// for example
+action.setIcon({
+  pathVariants: {
+    light: { 32: 'light-32.png' },
+    dark: { 32: 'dark-32.png' },
+  }
+});
+
+action.setIcon({
+  imageDataVariants: {
+    light: { 32: lightImageData },
+    dark: { 32: darkImageData },
+  }
+});
+```
+
+* Pros: Encapsules multiple ideas from this pull request's review comments.
+* Cons: Two new keys on setIcon() isn't necessarily better than one new key.
+
+---
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant is used when a dark color scheme is used for the
+surrounding UI, often referred to as "dark mode".
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+#### Optional or required?
+`icon_variants` is an optional key of the top level manifest dictionary.
+`icon_variants` is also an optional sub-key in the action key dictionary.
+
+#### Fallback
+* `icon_variants` will be used if they're supplied. Otherwise `default_icon`
+will be used if it's supplied. If neither are supplied, then `icons` will be
+used.
+
+* icon_variants or variants in action.setIcon() must provide both dark and light
+icon, if they occur.
+
+* Order of precedence for action:
+<api>.action.setIcon() -> action.icon_variants -> action.default_icon ->
+icon_variants -> icons
+
+* Order of precedence for non-action:
+icon_variants -> icons
+
+### FAQ
+* What happens if you specify icon_variants.dark but not icon_variants.light,
+and also icons?
+
+If either is provided, both are expected.
+
+* Do we use the dark mode icon from icon_variants for light as well, or fallback
+to icons?
+
+Both are expected if either "light" or "dark" are supplied.
+
+* What happens if you specify icon_variants.dark but not icon_variants.light, and
+you don't have icons?
+
+This could be a warning and a silent fallback to "icons", as the goal is to try
+to reduce errors as much as possible. Another option could be to fall back to
+the dark mode icon if in light mode and no light mode icon is supplied.
```

Comment:
All three of these seem to be the asking / answering the same thing slightly differently.

Reply:
This was my impression as well. A few thoughts:

- We should definitely support specifying light mode and not dark mode. Not all developers care about dark mode and having to supply both would be a burden.
- If you specify dark mode and not light mode, I have no strong feelings on if that is an error or not. I think "you must at least specify light mode" is a fair requirement.
- Falling back to `icons` from `icon_variants` seems nice in some ways but also introduces complexity, so I'm leaning towards saying we don't do that, but no strong feelings.

Reply:
If the developer only provides one icon, whether it's dark or light , it doesn't make much sense to use the new "icon_variants". In this case, they can use current "icon" and "action.default_icon" which already means for both dark and light icon. Therefore, the "icon_variants" is optional, if it is not provided, then fallback to others. But once provided, it is considered to represent all variant icons (i.e. if only provides one icon, it means both).

Reply:
@hanguokai `icon_variants` is quickly becoming a way to [specify alternate icon formats too](https://github.com/w3c/webextensions/pull/585#discussion_r1567170430), so having it only support one `color_scheme` for those extensions that don't care (or have universal icons that work anywhere) seems best.

I see `icon_variants` as a fixed version of `icons` and `default_icon` and it should be a viable option for any developer that wants to use the new thing.

Reply:
@xeenon A point of my [last comment](https://github.com/w3c/webextensions/pull/585#discussion_r1567403822) is that "once "icon_variants" is provided, it is considered to represent all variant icons (i.e. if only provides one icon, it means both).". What I mean is, if the user only provides one icon (`color_scheme` is dark or light, or no `color_scheme`), this icon is used for both dark and light icon. It doesn't fallback to "icon" or "action.default_icon".

Reply:
> I see icon_variants as a fixed version of icons and default_icon and it should be a viable option for any developer that wants to use the new thing.

Oh, I understand, you mean "icon_variants" is not just a supplement to "icons" or "default_icon", "icon_variants" can completely replace them. If developers provide "icon_variants", they don't need to provide "icons" anymore.

Reply:
Also leaning towards regarding `icon_variants` as a replacement for `icons` in the long run. just like `options_ui` has mostly replaced `options_page`. This will motivate extension developers to add more `icon_variants` as well for a better user experience.